Fix `pdflatex` and `pdftocairo` calls within LaTeX circuit drawer

adekusar-drl opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR fixes #6929. Currently there's an incorrect command for `pdflatex` in the code.


### Details and comments
The command `pdflatex -v` does not work on all operating systems, while `pdflatex -version` is an officially supported command, see `man pdflatex`.

Also `HAS_PDFLATEX` is not assigned to `True` if `pdflatex` is successed.

Catching of `SubprocessError` is required if a subprocess is called with additional checks like this `subprocess.run(..., checks=True)`, `CalledProcessError` which, in turns, is a sublclass of `SubprocessError`, may be raised.

jakelishman reviewed (dismissed):
Logically this change is sensible, but the fact that we weren't binding `HAS_PDFLATEX` if the command succeeded shows that the command _never_ succeeded, otherwise people would have seen import errors if they had `pdflatex` installed.

There's a few issues with the code still:
1. `subprocess.run("command -arg")` is trying to find a single command _named_ `command -arg` - it doesn't split on spaces.  It should be `subprocess.run(["command", "-arg"])`.
2. `pdflatex -version` dumps a bunch of information to `stdout` (or maybe `stderr` in MikTeX - I'm not 100% sure).  We don't want to spam the user, so we should set `stdout=subprocess.DEVNULL` and `stderr=subprocess.DEVNULL` as well.
3. Invoking `pdflatex -version` involves spawning a non-trivial subprocess, and it's currently unconditionally run on `import qiskit`.  We should lazily check this when it's actually queried for the first time, similar to how we handle `matplotlib` imports, to avoid the penalty (`qiskit` is already slower than we'd like to import).

All that said, on a closer look, no part of Terra actually _uses_ that import at all.  Do we even need this check?  We basically already test it lazily within `_latex_circuit_drawer`, which doesn't use the name test.

adekusar-drl commented:
> There's a few issues with the code still:
> 
>     1. `subprocess.run("command -arg")` is trying to find a single command _named_ `command -arg` - it doesn't split on spaces.  It should be `subprocess.run(["command", "-arg"])`.

Updated.
> 
>     2. `pdflatex -version` dumps a bunch of information to `stdout` (or maybe `stderr` in MikTeX - I'm not 100% sure).  We don't want to spam the user, so we should set `stdout=subprocess.DEVNULL` and `stderr=subprocess.DEVNULL` as well.

Agree, updated.

> 
>     3. Invoking `pdflatex -version` involves spawning a non-trivial subprocess, and it's currently unconditionally run on `import qiskit`.  We should lazily check this when it's actually queried for the first time, similar to how we handle `matplotlib` imports, to avoid the penalty (`qiskit` is already slower than we'd like to import).

I'm not the one who introduced this check for `pdflatex` and I have no idea why it is required, I'm trying to fix the issue I hit.

> All that said, on a closer look, no part of Terra actually _uses_ that import at all. Do we even need this check? We basically already test it lazily within `_latex_circuit_drawer`, which doesn't use the name test.

Actually there's an import, take a look at https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/visualization/__init__.py#L129
Without `HAS_PDFLATEX = True` there's another exception on that line: "cannot import name 'HAS_PDFLATEX'".

jakelishman commented:
Yeah, I saw that import, but that's just an `__init__` file: nothing else in Terra actually uses it from there.  The only place the string `HAS_PDFLATEX` is mentioned in Qiskit (at least Terra, Aer, Ignis, IMBQ, Optimization and the metapackage - the stuff I have cloned) is defining it (like in the import) or seeing if it exists in a test (so it's effectively been an unconditional skip on two test cases).

Since it was only introduced in #6487 a few days ago, I don't think there's going to be any trouble if we remove that whole test and the variable - looks like it got added in by mistake anyway.  We don't need to export the variable if `pdflatex` is only used in one place in the code, anyway.

jakelishman commented:
Ah ok, I suppose we _do_ need to have it so that test case can correctly be skipped.  Apologies, that's on me.

Since we do have to include it, we should make it so that `HAS_PDFLATEX` is lazily evaluated, like `HAS_MATPLOTLIB`.  For example: https://github.com/Qiskit/qiskit-terra/blob/b419e3968f7a88ff84d7968107c17ee226871513/qiskit/visualization/matplotlib.py#L1423-L1442

kdk commented:
> 3. Invoking pdflatex -version involves spawning a non-trivial subprocess, and it's currently unconditionally run on import qiskit. We should lazily check this when it's actually queried for the first time, similar to how we handle matplotlib imports, to avoid the penalty (qiskit is already slower than we'd like to import).

I agree this is a problem. Likely the `from qiskit.visualization import pass_manager_drawer` in `passmanager.py` should be a runtime import.

adekusar-drl commented:
I have updated `HAS_PDFLATEX` to be lazily evaluated as you suggested.

jakelishman commented on qiskit/visualization/circuit_visualization.py:
```diff
@@ -46,14 +46,34 @@
+class HasPdfLatexWrapper:
+    """Wrapper to lazily detect presence of the pdflatex command."""
+
+    has_pdflatex = False
+
+    # pylint: disable=unused-import
+    def __bool__(self):
+        if not self.has_pdflatex:
```
This will re-run the check every time the wrapper is queried (and the result is `False`).  We could do
```suggestion
    has_pdflatex = None

    def __bool__(self):
        if self.has_pdflatex is None:
```
to avoid that, and remove the pylint comment, since there won't be a warning generated here.

adekusar-drl replied:
Fixed, pylint is from a quick copy-paste, sorry.

jakelishman reviewed (commented):
Thanks for the change, it looks good (comment aside).  Please can we also make `_latex_circuit_drawer` query this wrapper as well now, so Terra only ever attempts to `pdflatex` once if the command fails?  We can raise `MissingOptionalLibraryError` (a Terra-specific error) if `HAS_PDFLATEX` is false, and still do the regular catching for non-zero return code.

1ucian0 commented:
@TharrmashasthaPV, do you mind also having a look here?

TharrmashasthaPV commented:
Nice work @adekusar-drl . But I think we have an issue here. Just having `pdflatex` does not guarantee that `qcircuit` package is also available. Say if a system has only the packages of `texlive-latex-base` installed, then `pdflatex -version` would give no error but when you try to latex draw a quantum circuit you will end up getting "Unable to compile latex" error. I understand we should have caught this in #6487 but somehow it slipped past us. So any thoughts on this?

adekusar-drl commented:
> 
> 
> Thanks for the change, it looks good (comment aside). Please can we also make `_latex_circuit_drawer` query this wrapper as well now, so Terra only ever attempts to `pdflatex` once if the command fails? We can raise `MissingOptionalLibraryError` (a Terra-specific error) if `HAS_PDFLATEX` is false, and still do the regular catching for non-zero return code.

Don't fully understand your idea here, but added a check to `_latex_circuit_drawer` if `HAS_PDFLATEX` is set to true. If not, the code raises `MissingOptionalLibraryError`.

adekusar-drl commented:
> 
> 
> Nice work @adekusar-drl . But I think we have an issue here. Just having `pdflatex` does not guarantee that `qcircuit` package is also available. Say if a system has only the packages of `texlive-latex-base` installed, then `pdflatex -version` would give no error but when you try to latex draw a quantum circuit you will end up getting "Unable to compile latex" error. I understand we should have caught this in #6487 but somehow it slipped past us. So any thoughts on this?

@TharrmashasthaPV Initially, it was a quick fix, just +3 lines, cause it was a major (imo) breakdown in the code. Now it is already +32 and I have no idea what you are talking about as I just don't know how drawing/plotting is implemented and what the dependencies are. So, if anything is missing, let me know.

jakelishman commented:
@adekusar-drl: if you'd like me to take over and finish off the last couple of comments, let me know - it's no trouble, and I can push them directly to this PR.

Sorry this ballooned a little bit.  The issue you pointed out was quite simple, but unfortunately fixing it popped up a whole slew of other stuff that had been lying dormant before because of the original bug you pointed out, and we can't merge a fix if we know it's uncovering more bugs.

adekusar-drl commented:
@jakelishman sure, go ahead, I really hope not that much left.

jakelishman commented:
@TharrmashasthaPV: I've taken over the PR and refactored the error handling to make it a bit more robust (and fixed a couple of places where the handling was buggy).

There's a limit to which we can detect an incomplete LaTeX environment ahead of time (say detecting a lack of `qcircuit.tex`); we _could_ do it with a dummy file, but it still wouldn't catch all cases of a badly formed environment, so we'll still always have to have the wrapper checks around the actual calls.  We certainly shouldn't do that within the `HAS_PDFLATEX` variable, since other later uses might not care about `qcircuit.tex`.  The worst-case scenario of it is just that we emit a `VisualizationError` instead of a `MissingOptionalLibraryError` (which were more meant for Python libraries anyway), which isn't a big deal.

jakelishman commented:
@1ucian0: since I'm now a large contributor to this PR, I'll pass off the reviewing/approval of it onto you.  GitHub will technically let me do it, but that would defeat the spirit of the process.

mtreinish commented on qiskit/visualization/__init__.py:
```diff
@@ -126,7 +126,7 @@
 from qiskit.visualization.array import array_to_latex
 
-from .circuit_visualization import circuit_drawer, HAS_PIL, HAS_PDFLATEX
+from .circuit_visualization import circuit_drawer, HAS_PIL, HAS_PDFLATEX, HAS_PDFTOCAIRO
```
Do we actually need to be exporting these? HAS_MATPLOTLIB is here mostly for backwards compatibility and not because I'd say it's something we wanted to support as part of the public interface

jakelishman replied:
I had the same comment about `HAS_PDFLATEX` originally, but then it transpires that it's useful to have these things for the tests (so they can decide whether they should skip or not).

1ucian0 commented:
Sure. Since @TharrmashasthaPV might have a more complete picture on the latex drawer, it would be great to have his opinion too.

1ucian0 commented:
LGTM. The @mtreinish comments from https://github.com/Qiskit/qiskit-terra/pull/6930/#discussion_r699530226 applies to all the `HAS_*` vars and it can be done in a follow up.
